python/node_similarity.py

```diff
@@ -0,0 +1,285 @@
+    """This method calculates the similarity of nodes for given method and mode.
+
+    :param node1: The first node or tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: The second node or tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param method: Similarity measure which will be used for calculating the similarity between nodes.
+    Currently available are `_calculate_jaccard`, `_calculate_overlap` and `_calculate_cosine`
+    :type method: function
+    :param mode: Can be `p` (or `pairwise`) for pairwise similarity calculation or `c` (or `cartesian`) for
+    calculating the similarity of the Cartesian product of given tuples
+    :type mode: str
+
+    :raises TypeError: Occurs if there's a type mismatch for passed arguments
+    :raises ValueError: Occurs if any passed argument is invalid
+
+    :return: Returns the calculated similarity between nodes in a list of mgp.Records. Each mgp.Record consist of
+    the first node, the second node and the similarity between them.
+    :rtype: list
+    """
+
+    nodes1: tuple
+    nodes2: tuple
+
+    if isinstance(node1, mgp.Vertex):
+        nodes1 = tuple([node1])
+    elif isinstance(node1, tuple):
+        nodes1 = node1
+    else:
+        raise TypeError("Invalid type of first argument.")
+
+    if isinstance(node2, mgp.Vertex):
+        nodes2 = tuple([node2])
+    elif isinstance(node2, tuple):
+        nodes2 = node2
+    else:
+        raise TypeError("Invalid type of second argument.")
+
+    result: list
+
+    if mode.lower() in Mode.PAIRWISE:
+        if len(nodes1) == len(nodes2):
+            return [
+                mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+                for n1, n2 in zip(nodes1, nodes2)
+            ]
+        else:
+            raise ValueError("Incompatible lengths of given arguments.")
+    elif mode.lower() in Mode.CARTESIAN:
+        return [
+            mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+            for n1, n2 in product(nodes1, nodes2)
+        ]
+    else:
+        raise ValueError("Invalid mode.")
+
+
+def _get_neighbors(node: mgp.Vertex) -> Set[mgp.Vertex]:
+    """This method finds all neighbors of a given node. If neighbors of a node have already been fetched once before,
+    they can be found in the neighbors_dict dictionary and thus the method returns the neighbors faster.
+    neighbors_dict is a global variable and it resets at the beginning of the program
+
+    :param node: A node
+    :type node: mgp.Vertex
+
+    :return: Set of all neighbors of a node
+    :rtype: set
+    """
+
+    if node.id in neighbors_dict:
+        return neighbors_dict[node.id]
+    else:
+        neighbors = set()
+
+        for edge in node.in_edges:
+            neighbors.add(edge.from_vertex)
+
+        for edge in node.out_edges:
+            neighbors.add(edge.to_vertex)
+
+        neighbors_dict[node.id] = neighbors
+
+        return neighbors
```

Comment:
If changes have been made to the graph since the previous query, this function wouldn’t reflect them because there’s no option to specify that the nodes’ neighbors are to be re-fetched.

Reply:
The dictionary `neighbors_dict` is reset at the beginning of each execution and all the neighbors are calculated anew.